src/typechecker.erl

```diff
@@ -3438,16 +3438,76 @@ refinable(_) ->
 no_guards({clause, _, _, Guards, _}) ->
     Guards == [].
 
+
+-spec check_guard_call(erlang:anno(), atom(), list()) -> map().
+check_guard_call(P, is_atom, [{var, _, Var}]) -> #{Var => {type, P, atom, []}};
+check_guard_call(P, is_binary, [{var, _, Var}]) -> #{Var => {type, P, binary, []}};
+check_guard_call(P, is_bitstring, [{var, _, Var}]) -> #{Var => {type, P, bitstring, []}};
+check_guard_call(P, is_boolean, [{var, _, Var}]) -> #{Var => {type, P, boolean, []}};
+check_guard_call(P, is_float, [{var, _, Var}]) -> #{Var => {type, P, float, []}};
+check_guard_call(P, is_function, [{var, _, Var}]) -> #{Var => {type, P, 'fun', []}};
+check_guard_call(P, is_function, [{var, _, Var}, _]) -> #{Var => {type, P, 'fun', []}};
+check_guard_call(P, is_integer, [{var, _, Var}]) -> #{Var => {type, P, integer, []}};
+check_guard_call(P, is_list, [{var, _, Var}]) -> #{Var => {type, P, list, []}};
+check_guard_call(P, is_map, [{var, _, Var}]) -> #{Var => {type, P, map, []}};
+check_guard_call(P, is_number, [{var, _, Var}]) -> #{Var => {type, P, number, []}};
+check_guard_call(P, is_pid, [{var, _, Var}]) -> #{Var => {type, P, pid, []}};
+check_guard_call(P, is_port, [{var, _, Var}]) -> #{Var => {type, P, port, []}};
+check_guard_call(P, is_record, [{var, _, Var}, {atom, _, Record}]) ->
+    #{Var => {type, P, record, [{atom, P, Record}]}};
+check_guard_call(P, is_record, [{var, _, Var}, {atom, _, Record}, _]) ->
+    #{Var => {type, P, record, [{atom, P, Record}]}};
+check_guard_call(P, is_reference, [{var, _, Var}]) -> #{Var => {type, P, reference, []}};
+check_guard_call(P, is_tuple, [{var, _, Var}]) -> #{Var => {type, P, tuple, []}};
+check_guard_call(_P, _Fun, _Vars) ->
+    #{}.
+
+-spec check_guard(#env{}, term()) -> map().
+check_guard(_Env, {call, P, {atom, _, Fun}, Vars}) ->
+    check_guard_call(P, Fun, Vars);
+check_guard(_Env, {call, P, {remote,_,_,{atom, _, Fun}}, Vars}) ->
+    check_guard_call(P, Fun, Vars);
```

Comment:
I assume the module must be `erlang` here. Better match it explicitly. We shouldn't accept `arbitrary_module:is_atom(X)`.

Reply:
So apparently, you can only have a few BIFs as guards. So there is no need to check for this because otherwise, erlang would complain at compilation and it wouldn't work. Hence why I removed the BIF check. http://erlang.org/doc/reference_manual/expressions.html#guard-sequences

Reply:
You're assuming that we only want to type check modules which compile.

It's fine with me, although we have lots code for checking code which doesn't compile. 

Anyway, withing `{atom, _, erlang}` here instead of `_` would make it easier to understand the code IMO.

Reply:
Do we still need to support the `remote` clause if we use erlang there?

Reply:
The pattern should be `{remote, _, {atom, _, erlang}, {atom, _, Fun}}`.

Reply:
If `erlang:is_atom(X)` is allowed in guards, we should better support it, because it would be represented as `remote` in the AST.

Reply:
It's not very important though, but since it's already implemented, better keep it.

Reply:
Done